ddtrace/tracer/writer.go

```diff
@@ -0,0 +1,277 @@
+		prioritySampling: s,
+	}
+}
+
+func (h *agentTraceWriter) add(trace []*span) {
+	if err := h.payload.push(trace); err != nil {
+		h.config.statsd.Incr("datadog.tracer.traces_dropped", []string{"reason:encoding_error"}, 1)
+		log.Error("error encoding msgpack: %v", err)
+	}
+	if h.payload.size() > payloadSizeLimit {
+		h.config.statsd.Incr("datadog.tracer.flush_triggered", []string{"reason:size"}, 1)
+		h.flush()
+	}
+}
+
+func (h *agentTraceWriter) stop() {
+	h.wg.Wait()
+}
+
+// flush will push any currently buffered traces to the server.
+func (h *agentTraceWriter) flush() {
+	if h.payload.itemCount() == 0 {
+		return
+	}
+	h.wg.Add(1)
+	h.climit <- struct{}{}
+	go func(p *payload) {
+		defer func(start time.Time) {
+			<-h.climit
+			h.wg.Done()
+			h.config.statsd.Timing("datadog.tracer.flush_duration", time.Since(start), nil, 1)
+		}(time.Now())
+		size, count := p.size(), p.itemCount()
+		log.Debug("Sending payload: size: %d traces: %d\n", size, count)
+		rc, err := h.config.transport.send(p)
+		if err != nil {
+			h.config.statsd.Count("datadog.tracer.traces_dropped", int64(count), []string{"reason:send_failed"}, 1)
+			log.Error("lost %d traces: %v", count, err)
+		} else {
+			h.config.statsd.Count("datadog.tracer.flush_bytes", int64(size), nil, 1)
+			h.config.statsd.Count("datadog.tracer.flush_traces", int64(count), nil, 1)
+			if err := h.prioritySampling.readRatesJSON(rc); err != nil {
+				h.config.statsd.Incr("datadog.tracer.decode_error", nil, 1)
+			}
+		}
+	}(h.payload)
+	h.payload = newPayload()
+}
+
+type logTraceWriter struct {
+	config    *config
+	buf       bytes.Buffer
+	hasTraces bool
+	w         io.Writer
+}
+
+var _ traceWriter = &logTraceWriter{}
+
+func newLogTraceWriter(c *config) *logTraceWriter {
+	w := &logTraceWriter{
+		config: c,
+		w:      os.Stdout,
+	}
+	w.resetPayload()
+	return w
+}
+
+const (
+	// maxFloatLength is the maximum length that a string encoded by encodeFloat can be.
+	// At 1e21 and beyond, floats are encoded to strings in exponential format. The longest
+	// a string encoded by encodeFloat can be is 22 characters.
+	maxFloatLength = 22
+
+	// suffix is the final string that the trace writer has to append to a payload to close
+	// the JSON.
+	logPayloadSuffix = "]}\n"
+
+	// logPayloadLimit is the maximum size log line allowed by cloudwatch
+	logPayloadLimit = 256 * 1024
+)
+
+func (h *logTraceWriter) resetPayload() {
+	h.buf.Reset()
+	h.buf.WriteString(`{"traces": [`)
+	h.hasTraces = false
+}
+
+// encodeFloat correctly encodes float64 to the format enforced by ES6
+func encodeFloat(p []byte, f float64) []byte {
+	if math.IsInf(f, -1) {
+		return append(p, "-Infinity"...)
+	} else if math.IsInf(f, 1) {
+		return append(p, "Infinity"...)
+	}
+	abs := math.Abs(f)
+	if abs != 0 && (abs < 1e-6 || abs >= 1e21) {
+		p = strconv.AppendFloat(p, f, 'e', -1, 64)
+		// clean up e-09 to e-9
+		n := len(p)
+		if n >= 4 && p[n-4] == 'e' && p[n-3] == '-' && p[n-2] == '0' {
```

Comment:
What is this wizardry? Why do we need it? Is there nothing in the stdlib that can help? Let's document it, and if it's copied from somewhere, let's mention where from.

Reply:
Added documentation about the magic 😄